sgm/inference/api.py

```diff
@@ -158,18 +160,51 @@ class SamplingSpec:
 class SamplingPipeline:
     def __init__(
         self,
-        model_id: ModelArchitecture,
-        model_path="checkpoints",
-        config_path="configs/inference",
+        model_id: Optional[ModelArchitecture] = None,
+        model_spec: Optional[SamplingSpec] = None,

```

Comment:
Could use ` | None` as well...

Reply:
From the rest of the code it also looks like the `spec` could be a string – did you run `mypy` or `pyright` over the typing?

Reply:
Please don't use `type | None`, since this will break in Python 3.8 which we want to support

Reply:
@jenuk It won't break if there's a `from __future__ import annotations` import.

Reply:
I'm on the fence about this after doing some more digging, `from __future__ import annotations` is not set to become a default soon, if ever: https://mail.python.org/archives/list/python-dev@python.org/message/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/

Reply:
It's widely used in the current Python ecosystem, though. 

There's no downside to using it unless you really do need to do runtime annotation analysis (like eg. Pydantic does), and even then it's not a problem.

One upside is much easier-to-read types. `ruff` (for which there's a PR...) can be used to auto-upgrade old-style types to new-style types later.